feat(onedrive): implemented secure file download functionality

amjedidiah opened the pull request:
- Added direct file download using Graph API and fetch
- Implemented progress tracking for downloads
- Added fallback mechanism for download URLs
- Handled file type and thumbnail preservation
- Added proper error handling and user feedback

BREAKING CHANGE: Changes the way OneDrive files are downloaded. Uses direct download URLs instead of sharing links


https://github.com/user-attachments/assets/09bf3ade-2549-493c-a170-6592e95b1ea7

amjedidiah committed:
feat(onedrive): implemented secure file download functionality

- Added direct file download using Graph API and fetch
- Implemented progress tracking for downloads
- Added fallback mechanism for download URLs
- Handled file type and thumbnail preservation
- Added proper error handling and user feedback

BREAKING CHANGE: Changes the way OneDrive files are downloaded. Uses direct download URLs instead of sharing links

amjedidiah committed:
fix(app): fixed bugs with onedrive authentication

BSalaeddin commented on src/hooks/useOneDriveAuth.ts:
```diff
@@ -15,84 +15,215 @@ import {
+                    expiresOn: response.expiresOn!.getTime(),
+                }
+                setToken(newToken)
+                localStorage.setItem(
```
MSAL's built-in caching mechanisms are likely a better alternative to localStorage for token management, as they are designed to securely handle tokens and minimize risks like XSS attacks. I'm not sure about the specific case here that led to using localStorage, but if acquireTokenSilent (which I see you've used elsewhere) works in this context, it would definitely be a safer and more reliable approach.

amjedidiah replied:
@BSalaeddin 

Not sure if we can do without either localStorage or sessionStorage
However, I will look this up. 


<img width="793" alt="Screenshot 2024-11-22 at 23 18 20" src="https://github.com/user-attachments/assets/586e6250-caed-48a5-b112-164fa250bc53">

BSalaeddin replied:
@amjedidiah , just to clarify, I wasn’t suggesting changing the entire cache location setting.
My thought was more about leveraging acquireTokenSilent to fetch the token directly without needing to manually set or retrieve it in localStorage during the flow. That said, I’m not entirely sure if it’s doable, as I haven’t worked with MSAL myself. I trust your judgment and appreciate you looking into it!

amjedidiah replied:
Thanks @BSalaeddin , done

BSalaeddin commented on src/components/UpupUploader/FileBrowser/BrowserOD.tsx:
```diff
@@ -14,59 +16,270 @@ type Props = {
+    ) => {
+        const response = await fetch(url, {
+            method: 'GET',
+            headers: { Accept: 'application/json' },
```
downloadFile function sets the Accept header to 'application/json' when fetching file content

amjedidiah replied:
Thanks @BSalaeddin, done

amjedidiah committed:
review(app): started review fixes

- removed unnecessary header in src/components/UpupUploader/FileBrowser/BrowserOD.tsx
- build fixes

amjedidiah committed:
refactor(auth): used MSAL built-in token caching instead of localStorage

- Removed manual token storage in localStorage in favour of acquireTokenSilent for token management
- Fixed logoutPopup type error
- Moved `pako` package from devDependencies to dependencies as it should be

MedAmine1212 commented on src/components/UpupUploader/FileBrowser/BrowserOD.tsx:
```diff
@@ -14,59 +16,269 @@ type Props = {
-            }) as unknown as OneDriveFile
+        const promises = files.map(async (file, index) => {
+            console.log('Processing file:', file)
+
```
Make sure to clean all the code from console.logs

amjedidiah replied:
Thanks @MedAmine1212 , done

MedAmine1212 commented on src/components/UpupUploader/FileBrowser/BrowserOD.tsx:
```diff
@@ -14,59 +16,269 @@ type Props = {
+                    throw new Error('Could not get download URL')
+                }
+
+                console.log('Got download URL:', downloadUrl)
```
remove console.log

amjedidiah replied:
Thanks @MedAmine1212 , done

amjedidiah committed:
chore(app): removed `console.log` statements

amjedidiah committed:
Merge branch 'master' into CU-86b06xn7j_feat-onedrive-file-download
